x/wasm/keeper/keeper.go

```diff
@@ -482,26 +484,17 @@ func (k Keeper) migrate(
 		contractInfo.IBCPortID = ibcPort
 	}
 
-	env := types.NewEnv(sdkCtx, contractAddress)
-
-	// prepare querier
-	querier := k.newQueryHandler(sdkCtx, contractAddress)
-
-	prefixStoreKey := types.GetContractStorePrefix(contractAddress)
-	vmStore := types.NewStoreAdapter(prefix.NewStore(runtime.KVStoreAdapter(k.storeService.OpenKVStore(sdkCtx)), prefixStoreKey))
-	gasLeft := k.runtimeGasForContract(sdkCtx)
-	res, gasUsed, err := k.wasmVM.Migrate(newCodeInfo.CodeHash, env, msg, vmStore, cosmwasmAPI, &querier, k.gasMeter(sdkCtx), gasLeft, costJSONDeserialization)
-	k.consumeRuntimeGas(sdkCtx, gasUsed)
-	if err != nil {
-		return nil, errorsmod.Wrap(types.ErrVMError, err.Error())
-	}
-	if res == nil {
-		// If this gets executed, that's a bug in wasmvm
-		return nil, errorsmod.Wrap(types.ErrVMError, "internal wasmvm error")
-	}
-	if res.Err != "" {
-		return nil, types.MarkErrorDeterministic(errorsmod.Wrap(types.ErrMigrationFailed, res.Err))
+	// check for migrate endpoint
+	hasMigrateEndpoint := false
+	switch {
+	case contains(report.Entrypoints, "migrate") && msg != nil:
+		hasMigrateEndpoint = true
+	case !contains(report.Entrypoints, "migrate") && msg == nil:
+		hasMigrateEndpoint = false
+	default:
+		return nil, errorsmod.Wrap(types.ErrMigrationFailed, "invalid migration setup")
```

Comment:
I think the error message could be a bit more descriptive. Maybe something like "invalid migration parameters". Still not great, but it's hard to word it in a way that describes both failure states ("no migrate + msg" and "migrate + no msg")